internal/platform/gin/auditlog/middleware.go

```diff
@@ -0,0 +1,188 @@
+
+func (fn optionFunc) apply(o *middlewareOptions) {
+	fn(o)
+}
+
+// WithClock sets the clock in an audit log middleware.
+func WithClock(clock Clock) Option {
+	return optionFunc(func(o *middlewareOptions) {
+		o.clock = clock
+	})
+}
+
+// WithSensitivePaths marks API call paths as sensitive, causing the log entry to omit the request body.
+func WithSensitivePaths(sensitivePaths []*regexp.Regexp) Option {
+	return optionFunc(func(o *middlewareOptions) {
+		o.sensitivePaths = sensitivePaths
+	})
+}
+
+// WithErrorHandler sets the clock in an audit log middleware.
+func WithErrorHandler(errorHandler ErrorHandler) Option {
+	return optionFunc(func(o *middlewareOptions) {
+		o.errorHandler = errorHandler
+	})
+}
+
+// WithUserIDExtractor sets the function that extracts the user ID from the request.
+func WithUserIDExtractor(userIDExtractor func(req *http.Request) uint) Option {
+	return optionFunc(func(o *middlewareOptions) {
+		o.userIDExtractor = userIDExtractor
+	})
+}
+
+// Middleware returns a new HTTP middleware that records audit log entries.
+func Middleware(driver Driver, opts ...Option) gin.HandlerFunc {
+	options := middlewareOptions{
+		clock:           realClock{},
+		userIDExtractor: func(req *http.Request) uint { return 0 },
+		errorHandler:    NoopErrorHandler{},
+	}
+
+	for _, opt := range opts {
+		opt.apply(&options)
+	}
+
+	return func(c *gin.Context) {
+		path := c.Request.URL.Path
+
+		if c.Request.URL.RawQuery != "" {
+			path = path + "?" + c.Request.URL.RawQuery
+		}
+
+		entry := Entry{
+			Time:          options.clock.Now(),
+			CorrelationID: c.GetString(correlationid.ContextKey),
+			HTTP: HTTPEntry{
+				ClientIP:  c.ClientIP(),
+				UserAgent: c.Request.UserAgent(),
+				Method:    c.Request.Method,
+				Path:      path,
+			},
+		}
+
+		var sensitiveCall bool
+
+		// Determine if this call contains sensitive information in its request body.
+		for _, r := range options.sensitivePaths {
+			if r.MatchString(c.Request.URL.Path) {
+				sensitiveCall = true
+				break
+			}
+		}
+
+		// Only override the request body if there is actually one and it doesn't contain sensitive information.
+		saveBody := c.Request.Body != nil && !sensitiveCall
+
+		var buf bytes.Buffer
+
+		if saveBody {
+			// This should be ok, because the server keeps a reference to the original body,
+			// so it can close the original request itself.
+			c.Request.Body = ioutil.NopCloser(io.TeeReader(c.Request.Body, &buf))
+		}
+
+		c.Next() // process request
+
+		entry.UserID = options.userIDExtractor(c.Request)
+
+		// Consider making this configurable if you need to log unauthorized requests,
+		// but keep in mind that in case of a public installation it's a potential DoS attack vector.
+		if c.Writer.Status() == http.StatusUnauthorized {
+			return
+		}
+
+		entry.HTTP.StatusCode = c.Writer.Status()
+		entry.HTTP.ResponseSize = c.Writer.Size()
+		entry.HTTP.ResponseTime = int(options.clock.Since(entry.Time).Nanoseconds() / 1000 / 1000) // ms
```

Comment:
**Question**: what is the reason behind using `time.Duration{}.Nanoseconds() / 1000 / 1000` manual division instead of `time.Duration{}.Milliseconds()`?

Reply:
Dunno, copy-pasta. Will change it.